File: 1/1-16.ts
function fastExptIter(b: number, n: number, a: number): number {
    // Iterative exponentiation using successive squaring
    if (n === 0) {
        return a;
    } else if (n % 2 === 0) {
        return fastExptIter(b * b, Math.floor(n / 2), a);
    } else {
        return fastExptIter(b, n - 1, a * b);
    }
}

function fastExpt(b: number, n: number): number {
    // Compute b^n using logarithmic number of steps
    return fastExptIter(b, n, 1);
}

const base = 2;
const exponent = 10;
const result = fastExpt(base, exponent);
console.log(`${base}^${exponent} = ${result}`);

File: 1/1-16.rs
fn fast_expt_iter(b: f64, n: i32, a: f64) -> f64 {
    // Iterative exponentiation using successive squaring
    if n == 0 {
        a
    } else if n % 2 == 0 {
        fast_expt_iter(b * b, n / 2, a)
    } else {
        fast_expt_iter(b, n - 1, a * b)
    }
}

fn fast_expt(b: f64, n: i32) -> f64 {
    // Compute b^n using logarithmic number of steps
    fast_expt_iter(b, n, 1.0)
}

fn main() {
    let base = 2.0;
    let exponent = 10;
    let result = fast_expt(base, exponent);
    println!("{}^{} = {}", base, exponent, result);
}